### PR TITLE
[BugFix] Use MergedCommitID for merged pulls to create Diff/Patch File

### LINKED ISSUE #10932: [Bug] 500 on download pull-diff/patch file of a merged with deleted HeadBranch
test: https://gitea.com/gitnex/GitNex/pulls/340.diff

Gitea Version: 1.12.0+dev-82-g3723b0647

### PR DESCRIPTION
**Pull request by 6543**
fix #10932

and also fix "Empty Diff/Patch File when pull is merged"

### CODE REVIEW

File: routers/repo/pull.go [hunk 1]
```diff
@@ -1162,6 +1162,22 @@ func DownloadPullDiffOrPatch(ctx *context.Context, patch bool) {
 
 	pr := issue.PullRequest
 
+	if pr.HasMerged {
+		diffType := git.RawDiffType("diff")
+		if patch {
+			diffType = git.RawDiffType("patch")
```
Comment (jolheiser) [inline routers/repo/pull.go]: These should use the constants.
https://github.com/go-gitea/gitea/blob/3723b0647f61717a249f2379b1fed1e6b4dbdbce/modules/git/diff.go#L24-L28

**Review by jolheiser — CHANGES_REQUESTED** [review]

File: routers/repo/pull.go (hunk 1 above)
Comment (6543) [reply routers/repo/pull.go]: @jolheiser done

File: routers/repo/pull.go [hunk 2]
```diff
@@ -1162,6 +1162,22 @@ func DownloadPullDiffOrPatch(ctx *context.Context, patch bool) {
...
+		}
+		if err := git.GetRawDiff(
+			models.RepoPath(ctx.Repo.Owner.Name, ctx.Repo.Repository.Name),
+			pr.MergedCommitID,
+			diffType,
+			ctx.Resp,
+		); err != nil {
+			ctx.ServerError("GetRawDiff", err)
+			return
+		}
```
Comment (jolheiser) [inline routers/repo/pull.go]: This will need a final return if it's successful, otherwise it will continue.

File: routers/repo/pull.go [hunk 3]
```diff
@@ -1162,6 +1162,22 @@ func DownloadPullDiffOrPatch(ctx *context.Context, patch bool) {
...
+	if pr.HasMerged {
+		diffType := git.RawDiffNormal
+		if patch {
+			diffType = git.RawDiffPatch
+		}
+		if err := git.GetRawDiff(
+			models.RepoPath(ctx.Repo.Owner.Name, ctx.Repo.Repository.Name),
+			pr.MergedCommitID,
+			diffType,
+			ctx.Resp,
```
Comment (jolheiser) [inline routers/repo/pull.go]: Down below we use `ctx` to fulfill `io.Writer`, so why `ctx.Resp` here? It seems both work for the interface, but should they be consistent or is there a specific reason to use `ctx.Resp`?

File: routers/repo/pull.go [hunk 4]
```diff
@@ -1162,6 +1162,22 @@ func DownloadPullDiffOrPatch(ctx *context.Context, patch bool) {
...
+		); err != nil {
+			ctx.ServerError("GetRawDiff", err)
+			return
+		}
+	}
```
Comment (6543) [inline routers/repo/pull.go:1180]:
```suggestion
	return
	}
```

File: routers/repo/pull.go (hunk 3 above)
Comment (6543) [reply routers/repo/pull.go]: no this call was copyed from code witch serve commit diff/patch

Comment (6543) [inline routers/repo/pull.go]:
```suggestion
			ctx,
```

**Comment by 6543** [comment]
ToDo: add a test

**Comment by zeripath** [comment]
So actually I wonder if we should just simplify `services/pull/patch.go:23:func DownloadDiffOrPatch(...)` to:

```go
// DownloadDiffOrPatch will write the patch for the pr to the writer
func DownloadDiffOrPatch(pr *models.PullRequest, w io.Writer, patch bool) error {
	if err := pr.LoadBaseRepo(); err != nil {
		log.Error("Unable to load base repository ID %d for pr #%d [%d]", pr.BaseRepoID, pr.Index, pr.ID)
		return err
	}

	gitRepo, err := git.OpenRepository(pr.BaseRepo.RepoPath())
	if err != nil {
		return fmt.Errorf("OpenRepository: %v", err)
	}
	defer gitRepo.Close()
	if err := gitRepo.GetDiffOrPatch(pr.MergeBase, pr.GetGitRefName(), w, patch); err != nil {
		log.Error("Unable to get patch file from %s to %s in %s Error: %v", pr.MergeBase, pr.HeadBranch, pr.BaseRepo.FullName(), err)
		return fmt.Errorf("Unable to get patch file from %s to %s in %s Error: %v", pr.MergeBase, pr.HeadBranch, pr.BaseRepo.FullName(), err)
	}
	return nil
}
```

E.g. the patch is:

```diff
diff --git a/services/pull/patch.go b/services/pull/patch.go
index 96145fcd9..776bfdb75 100644
--- a/services/pull/patch.go
+++ b/services/pull/patch.go
@@ -21,30 +21,17 @@ import (
 
 // DownloadDiffOrPatch will write the patch for the pr to the writer
 func DownloadDiffOrPatch(pr *models.PullRequest, w io.Writer, patch bool) error {
-       // Clone base repo.
-       tmpBasePath, err := createTemporaryRepo(pr)
-       if err != nil {
-               log.Error("CreateTemporaryPath: %v", err)
+       if err := pr.LoadBaseRepo(); err != nil {
+               log.Error("Unable to load base repository ID %d for pr #%d [%d]", pr.BaseRepoID, pr.Index
                return err
        }
-       defer func() {
-               if err := models.RemoveTemporaryPath(tmpBasePath); err != nil {
-                       log.Error("DownloadDiff: RemoveTemporaryPath: %s", err)
-               }
-       }()
 
-       gitRepo, err := git.OpenRepository(tmpBasePath)
+       gitRepo, err := git.OpenRepository(pr.BaseRepo.RepoPath())
        if err != nil {
                return fmt.Errorf("OpenRepository: %v", err)
        }
        defer gitRepo.Close()
-
-       pr.MergeBase, err = git.NewCommand("merge-base", "--", "base", "tracking").RunInDir(tmpBasePath)
-       if err != nil {
-               pr.MergeBase = "base"
-       }
-       pr.MergeBase = strings.TrimSpace(pr.MergeBase)
-       if err := gitRepo.GetDiffOrPatch(pr.MergeBase, "tracking", w, patch); err != nil {
+       if err := gitRepo.GetDiffOrPatch(pr.MergeBase, pr.GetGitRefName(), w, patch); err != nil {
                log.Error("Unable to get patch file from %s to %s in %s Error: %v", pr.MergeBase, pr.Head
                return fmt.Errorf("Unable to get patch file from %s to %s in %s Error: %v", pr.MergeBase,
        }
```

**Comment by zeripath** [comment]
Otherwise, if that's too much - i.e. we use DownloadDiffOrPatch before the head is updated, we should make `DownloadDiffOrPatch` :

```go
// DownloadDiffOrPatch will write the patch for the pr to the writer
func DownloadDiffOrPatch(pr *models.PullRequest, w io.Writer, patch bool) error {
	if pr.HasMerged {
		if err := pr.LoadBaseRepo(); err != nil {
			log.Error("Unable to load base repository ID %d for pr #%d [%d]", pr.BaseRepoID, pr.Index, pr.ID)
			return err
		}

		gitRepo, err := git.OpenRepository(pr.BaseRepo.RepoPath())
		if err != nil {
			return fmt.Errorf("OpenRepository: %v", err)
		}
		defer gitRepo.Close()
		if err := gitRepo.GetDiffOrPatch(pr.MergeBase, pr.GetGitRefName(), w, patch); err != nil {
			log.Error("Unable to get patch file from %s to %s in %s Error: %v", pr.MergeBase, pr.HeadBranch, pr.BaseRepo.FullName(), err)
			return fmt.Errorf("Unable to get patch file from %s to %s in %s Error: %v", pr.MergeBase, pr.HeadBranch, pr.BaseRepo.FullName(), err)
		}
		return nil
	}

	// Clone base repo.
	tmpBasePath, err := createTemporaryRepo(pr)
	if err != nil {
		log.Error("CreateTemporaryPath: %v", err)
		return err
	}
	defer func() {
		if err := models.RemoveTemporaryPath(tmpBasePath); err != nil {
			log.Error("DownloadDiff: RemoveTemporaryPath: %s", err)
		}
	}()

	gitRepo, err := git.OpenRepository(tmpBasePath)
	if err != nil {
		return fmt.Errorf("OpenRepository: %v", err)
	}
	defer gitRepo.Close()

	pr.MergeBase, err = git.NewCommand("merge-base", "--", "base", "tracking").RunInDir(tmpBasePath)
	if err != nil {
		pr.MergeBase = "base"
	}
	pr.MergeBase = strings.TrimSpace(pr.MergeBase)
	if err := gitRepo.GetDiffOrPatch(pr.MergeBase, "tracking", w, patch); err != nil {
		log.Error("Unable to get patch file from %s to %s in %s Error: %v", pr.MergeBase, pr.HeadBranch, pr.BaseRepo.FullName(), err)
		return fmt.Errorf("Unable to get patch file from %s to %s in %s Error: %v", pr.MergeBase, pr.HeadBranch, pr.BaseRepo.FullName(), err)
	}
	return nil
}
```

Then you can drop your changes in `routers/repo/pull.go`.

**Comment by zeripath** [comment]
As we don't use DownloadDiffOrPatch elsewhere - I've uploaded a different PR that should solve the issue more cleanly.

**Comment by 6543** [comment]
close in faifor of #10936

**Pull request closed without merging**